### PR TITLE
Store only result hash in storage

### LINKED ISSUE #1020: Expired data request should have status Expired
Currently it will just be ignored

### LINKED ISSUE #1232: Store only result hash in storage
- Hash all information of result (data, requested time, resolved time, expiration time) and store this value in storage instead
- Emit result detail as an event.

Let's make sure we re-use the same result packet for this.

### PR DESCRIPTION
**Pull request by Benzbeeb**
- [x] Hash all information ```H(H(reqPacket)||H(resPacket))``` and store this value in storage instead

 

- [x] Emit result detail as an event.

Fixed #1232 
Fixed #1020

### CODE REVIEW

File: chain/x/oracle/keeper/request.go [hunk 1]
```diff
@@ -88,43 +93,51 @@ func (k Keeper) ProcessOracleResponse(
...
-	// 		sequence, request.SourcePort, request.SourceChannel, destinationPort, destinationChannel,
-	// 		1000000000, // Arbitrarily high timeout for now
-	// 	))
-
-	// 	if err != nil {
-	// 		fmt.Println("SEND PACKET ERROR", err)
-	// 	}
+	reqPacketData := types.NewOracleRequestPacketData(
+		request.ClientID, request.OracleScriptID,
+		string(request.Calldata), request.SufficientValidatorCount,
```
Comment (taobun) [inline chain/x/oracle/keeper/request.go]: hex.EncodeToString(request.Calldata)

File: chain/x/oracle/keeper/request.go [hunk 2]
```diff
@@ -88,43 +93,51 @@ func (k Keeper) ProcessOracleResponse(
...
+		request.RequestTime, ctx.BlockTime().Unix(),
+		types.Success, hex.EncodeToString(result),
+	)
+
+	k.AddResult(ctx, reqID, reqPacketData, resPacketData)
+
+	sourceChannelEnd, found := k.ChannelKeeper.GetChannel(ctx, request.SourcePort, request.SourceChannel)
+	if !found {
+		fmt.Println("SOURCE NOT FOUND1", request.SourcePort, request.SourceChannel)
+		return
```
Comment (taobun) [inline chain/x/oracle/keeper/request.go:151]: Choose one

File: chain/x/oracle/keeper/result.go [hunk 3]
```diff
@@ -1,68 +1,62 @@
...
 func (k Keeper) AddResult(
-	ctx sdk.Context, requestID types.RequestID, oracleScriptID types.OracleScriptID,
-	calldata []byte, result []byte,
+	ctx sdk.Context, requestID types.RequestID,
+	requestPacket types.OracleRequestPacketData,
+	responsePacket types.OracleResponsePacketData,
 ) error {
-	if uint64(len(result)) > k.GetParam(ctx, types.KeyMaxResultSize) {
+
+	if uint64(len([]byte(responsePacket.Result))) > k.GetParam(ctx, types.KeyMaxResultSize) {
```
Comment (taobun) [inline chain/x/oracle/keeper/result.go]: Seem weird

**Review by taobun — CHANGES_REQUESTED** [review]

File: chain/db/db.go [hunk 4]
```diff
@@ -570,10 +570,39 @@ func (b *BandDB) GetInvolvedAccountsFromTransferEvents(logs sdk.ABCIMessageLogs)
 	return involvedAccounts
 }
 
-func (b *BandDB) ResolveRequest(id int64, resolveStatus oracle.ResolveStatus, result []byte) error {
+func (b *BandDB) ResolveRequest(kvMap map[string]string) error {
```
Comment (taobun) [inline chain/db/db.go]: Let's move this function to `endblock.go` (this file is too big)

File: chain/x/oracle/keeper/request.go [hunk 5]
```diff
@@ -79,52 +84,116 @@ func (k Keeper) AddRequest(
 	return requestID, nil
 }
 
+func (k Keeper) handleResolveRequest(
```
Comment (taobun) [inline chain/x/oracle/keeper/request.go]: Rename to resolveRequest

File: chain/db/db.go [hunk 6]
```diff
@@ -570,10 +570,39 @@ func (b *BandDB) GetInvolvedAccountsFromTransferEvents(logs sdk.ABCIMessageLogs)
...
+		panic(err)
+	}
+
+	numResolveStatus, err := strconv.ParseInt(kvMap[oracle.AttributeKeyResolveStatus], 10, 8)
+	if err != nil {
+		panic(err)
+	}
+
+	resolveStatus := oracle.ResolveStatus(numResolveStatus)
 	if resolveStatus == 1 {
```
Comment (taobun) [inline chain/db/db.go]: Let's add all resolve status to alias and compare with meaningful status

File: chain/x/oracle/keeper/request.go [hunk 7]
```diff
@@ -79,52 +84,116 @@ func (k Keeper) AddRequest(
...
+		return
+	}
+
+	request, err := k.GetRequest(ctx, reqID)
+	if err != nil {
+		ctx.EventManager().EmitEvents(sdk.Events{
+			sdk.NewEvent(
+				types.EventTypeRequestExecute,
+				sdk.NewAttribute(types.AttributeKeyRequestID, fmt.Sprintf("%d", reqID)),
+				sdk.NewAttribute(types.AttributeKeyResolveStatus, fmt.Sprintf("%d", resolveStatus)),
```
Comment (taobun) [inline chain/x/oracle/keeper/request.go]: It will emit resolve completed without result?

File: chain/x/oracle/keeper/result.go [hunk 8]
```diff
@@ -1,68 +1,66 @@
...
 ) error {
+	result, err := hex.DecodeString(responsePacket.Result)
+	if err != nil {
+		return err
+	}
 	if uint64(len(result)) > k.GetParam(ctx, types.KeyMaxResultSize) {
 		return sdkerrors.Wrapf(types.ErrBadDataValue,
 			"AddResult: Result size (%d) exceeds the maximum size (%d).",
-			len(result), k.GetParam(ctx, types.KeyMaxResultSize),
+			len([]byte(responsePacket.Result)), k.GetParam(ctx, types.KeyMaxResultSize),
```
Comment (taobun) [inline chain/x/oracle/keeper/result.go]: You compare with length of result?

File: chain/x/oracle/keeper/request.go [hunk 9]
```diff
@@ -79,52 +84,116 @@ func (k Keeper) AddRequest(
...
+			sdk.NewAttribute(types.AttributeKeyResolvedTime, fmt.Sprintf("%d", resPacketData.ResolveTime)),
+			sdk.NewAttribute(types.AttributrKeyExpirationHeight, fmt.Sprintf("%d", request.ExpirationHeight)),
+		)})
+}
+
 // ProcessOracleResponse takes a
 func (k Keeper) ProcessOracleResponse(
 	ctx sdk.Context, reqID types.RequestID, resolveStatus types.ResolveStatus, result []byte,
 ) {
+	k.handleResolveRequest(ctx, reqID, resolveStatus, result)
```
Comment (taobun) [inline chain/x/oracle/keeper/request.go]: What does this function actually do?

File: chain/x/oracle/keeper/request.go [hunk 10]
```diff
@@ -79,52 +84,116 @@ func (k Keeper) AddRequest(
...
-	// 		1000000000, // Arbitrarily high timeout for now
-	// 	))
-
-	// 	if err != nil {
-	// 		fmt.Println("SEND PACKET ERROR", err)
-	// 	}
+		return
+	}
+
+	resPacketData := types.NewOracleResponsePacketData(
```
Comment (taobun) [inline chain/x/oracle/keeper/request.go]: It seems to create this packet 2 times for no reason.

**Review by taobun — CHANGES_REQUESTED** [review]

File: chain/x/oracle/keeper/request.go (hunk 5 above)
Comment (taobun) [reply chain/x/oracle/keeper/request.go]: Why resolve this comment without doing anything?

File: chain/x/oracle/keeper/request.go [hunk 11]
```diff
@@ -73,49 +78,100 @@ func (k Keeper) AddRequest(ctx sdk.Context, req types.Request) (types.RequestID,
 	return id, nil
 }
 
+func (k Keeper) resolveRequest(
+	ctx sdk.Context, reqID types.RequestID, resolveStatus types.ResolveStatus, result []byte,
+) types.OracleResponsePacketData {
+
+	request, err := k.GetRequest(ctx, reqID)
```
Comment (sorawit) [inline chain/x/oracle/keeper/request.go]: MustGetRequest

**Review by sorawit — CHANGES_REQUESTED** [review]

File: chain/db/table.go [hunk 12]
```diff
@@ -133,6 +133,9 @@ type Request struct {
 	TxHash                   []byte `gorm:"not null"`
 	ClientID                 string `gorm:"not null"`
 	Result                   []byte `sql:"default:null"`
+	RequestTime              int64  `sql:"default:null"`
+	ResolveTime              int64  `sql:"default:null"`
+	ExpirationTime           int64  `sql:"default:null"`
```
Comment (taobun) [inline chain/db/table.go]: ExpirationHeight

**Review by taobun — CHANGES_REQUESTED** [review]

File: chain/x/oracle/keeper/request.go [hunk 13]
```diff
@@ -73,49 +78,100 @@ func (k Keeper) AddRequest(ctx sdk.Context, req types.Request) (types.RequestID,
...
+	}
+	ctx.EventManager().EmitEvents(sdk.Events{
+		sdk.NewEvent(
+			types.EventTypeRequestExecute,
+			sdk.NewAttribute(types.AttributeKeyRequestID, fmt.Sprintf("%d", reqID)),
+			sdk.NewAttribute(types.AttributeKeyResolveStatus, fmt.Sprintf("%d", types.Success)),
+			sdk.NewAttribute(types.AttributeKeyResult, fmt.Sprintf("%s", hex.EncodeToString(result))),
+			sdk.NewAttribute(types.AttributeKeyRequestTime, fmt.Sprintf("%d", request.RequestTime)),
+			sdk.NewAttribute(types.AttributeKeyResolvedTime, fmt.Sprintf("%d", resPacketData.ResolveTime)),
+			sdk.NewAttribute(types.AttributrKeyExpirationHeight, fmt.Sprintf("%d", 10)), // TODO: REMOVE THIS
```
Comment (taobun) [inline chain/x/oracle/keeper/request.go]: currentBlock + 10

**Review by taobun — CHANGES_REQUESTED** [review]

**Review by sorawit — DISMISSED** [review]

**Review by taobun — DISMISSED** [review]

**Review by taobun — APPROVED** [review]

**Review by sorawit — APPROVED** [review]

**Pull request merged**